MAINT: travis-ci: Update CI scripts.

WarrenWeckesser opened the pull request:
MAINT: travis-ci: Update CI scripts.

* Use `set -ex` in tools/travis-before-install.sh.
* Install build dependencies in tools/travis-before-install.sh.
* Update a comment about an additional option in CFLAGS when USE_DEBUG=1.
* Clear PYTHONOPTIMIZE when running `pip install -r test_requirements.txt`
  because version 2.19 of pycparser (a dependency of one of the packages
  in test_requirements.txt) does not provide a wheel, and the source tar
  file does not install correctly when Python's optimization level is set
  to strip docstrings (see https://github.com/eliben/pycparser/issues/291).
* Remove unnecessary calls of `pip install` in tools/travis-test.sh.

Closes gh-15108.

seberg commented on tools/travis-before-install.sh:
```diff
@@ -33,5 +33,9 @@ python -V
+# pycparser doesn't work with optimized Python builds that
+# strip out docstrings, so preemptively install pycparser
+# with PYTHONOPTIMIZE unset.
+PYTHONOPTIMIZE= pip install -v pycparser
```
```suggestion
PYTHONOPTIMIZE="" pip install -v pycparser
```
I am daft when it comes to bash. Can you ass the `set -e` on the top as well? If this is fixed in newer pycparser versions, maybe you can change comment to reflect that it will be OK to delete the line again?

mattip commented on tools/travis-before-install.sh:
```diff
@@ -33,5 +36,11 @@ python -V
+# set to strip docstrings (see https://github.com/eliben/pycparser/issues/291),
+# so preemptively install pycparser with PYTHONOPTIMIZE unset.  If a future
+# version of pycparser doesn't have this problem, the following line can be
+# removed.
```
I think the comment should mention that pycparser does not ship a wheel, so it must be built and building fails with PYTHONOPTIMIZE=2

mattip commented on tools/travis-before-install.sh:
```diff
@@ -33,5 +36,11 @@ python -V
+# version of pycparser doesn't have this problem, the following line can be
+# removed.
+PYTHONOPTIMIZE="" pip install -v pycparser
 pip install -r test_requirements.txt
```
Why not combine this with the next line `PYTHONOPTIMIZE= pip install -r test_requirements.txt`

WarrenWeckesser replied:
That turns off the optimize flags for all the packages in `test_requirements.txt`.  I guess that would OK.  Presumably the intent of using `PYTHONOPTIMIZE=2` is to test *numpy* when built with that setting, so we don't care how the packages in `test_requirements.txt` are built.

WarrenWeckesser commented:
Updated with @mattip's suggestion to use `PYTHONOPTIMIZE="" pip install -r test_requirements.txt` instead of installing `pycparser` separately.

The shippable failure is not related to the PR.

mattip commented:
LGTM. Line 381 in the [relevant build log](https://travis-ci.org/numpy/numpy/jobs/633960731?utm_medium=notification&utm_source=github_status) is now "clean" where previously it erred
```
Processing /home/travis/.cache/pip/wheels/f2/9a/90/de94f8556265ddc9d9c8b271b0f63e57b26fb1d67a45564511/pycparser-2.19-py2.py3-none-any.whl
```
But later on I see in line 4826 we again call `pip install -r test_requirements.txt`. I think this is redundant. It comes from [toosl/travis-test.sh](https://github.com/numpy/numpy/blob/v1.18.1/tools/travis-test.sh#L68). Should we remove it as part of this PR or in a separate one?

eric-wieser commented on tools/travis-before-install.sh:
```diff
@@ -36,12 +36,5 @@ python -V
 popd
 
 pip install --upgrade pip setuptools
```
tools/travis-test.sh:145 seems to install `setuptools` a second time, unless I'm missing something

WarrenWeckesser replied:
@eric-wieser, yes, I've noticed that the travis-ci scripts don't follow DRY very well.  `wheel` and `setuptools` are installed here and there.  Also, `cython` is in `test_requirement.txt`, but in fact `cython` is a build requirement.  I'm looking into this, so consider this PR to be WIP.

WarrenWeckesser replied:
I removed the redundant calls to `pip install` in the latest update.

eric-wieser commented on tools/travis-before-install.sh:
```diff
@@ -33,5 +36,6 @@ python -V
 
 pip install --upgrade pip setuptools
-pip install -r test_requirements.txt
+pip install Cython==0.29.14
```
We need to keep this in a text file in order for @dependabot to find it I think. Perhaps we need a `build_requires.txt`, which could also include a setuptools version?

WarrenWeckesser replied:
The `requires` section in `pyproject.toml` lists the build requirements.  I just now getting up to speed on which tools actually use that file.  It would be nice if we could have that list in just one place.  (Python packaging... $%@#@#^$$*@#$!)

eric-wieser commented:
Do you want to split this PR into the bit that sets the optimize flag so that we can merge that, and the bit trying to clean up all the repeated installations?

WarrenWeckesser commented:
@eric-wieser, let's see if my latest update looks acceptable.  If so, we can keep this in one PR.

WarrenWeckesser commented:
All green with the latest update.  The scripts now installs the build dependencies from `pyproject.toml`, and the redundant calls to `pip install` have been removed.

rgommers commented on tools/travis-test.sh:
```diff
@@ -48,7 +48,10 @@ setup_base()
-    # Python3.5-dbg on travis seems to need this
+    pip install --upgrade setuptools toml
+    # Install the build dependencies from pyproject.toml.
+    pip install `python -c "import toml; req = toml.load('pyproject.toml')['build-system']['requires']; print(' '.join(req))"`
```
This is quite weird, overall it's less readable. Why do this?

WarrenWeckesser replied:
It installs the build dependencies from `pyproject.toml`.  That's where the list of dependencies now resides, correct?  (See https://github.com/numpy/numpy/pull/15275#discussion_r364186088)

It turns out I'm still iterating on this PR, so expect this to change shortly.  I'll still use this code snippet, but not here.

mattip replied:
pyproject.toml, which is a thing still in flux since there is not even an official toml reader in the stdlib, holds the build dependencies. The test_requirements.txt holds the test dependencies. I think for CI we should stick with the test dependencies and not mess with the toml file.

mattip replied:
using only the pyproject.toml will break dependabot.

WarrenWeckesser replied:
Which files does dependabot read?

WarrenWeckesser replied:
Partially answering my own question: dependabot looks at `test_requirements.txt`.  Does it read anything else?

WarrenWeckesser committed:
MAINT: travis-ci: Update CI scripts.

* Use `set -ex` in tools/travis-before-install.sh.
* Install build dependencies in tools/travis-before-install.sh.
* Update a comment about an additional option in CFLAGS when USE_DEBUG=1.
* Clear PYTHONOPTIMIZE when running `pip install -r test_requirements.txt`
  because version 2.19 of pycparser (a dependency of one of the packages
  in test_requirements.txt) does not provide a wheel, and the source tar
  file does not install correctly when Python's optimization level is set
  to strip docstrings (see https://github.com/eliben/pycparser/issues/291).
* Remove unnecessary calls of `pip install` in tools/travis-test.sh.

Closes gh-15108.

WarrenWeckesser commented:
Updated with a simpler set of changes.

mattip commented on tools/travis-before-install.sh:
```diff
@@ -32,6 +36,16 @@ python -V
+# We get the argument for 'cython' from test_requirements.txt to avoid
+# duplicating the version requirement stored there.
+pip install setuptools wheel `grep cython test_requirements.txt`
+
```
Which run fails without this line?

WarrenWeckesser replied:
One or more of the `setup.py` installs... I think.  Let's find out...

WarrenWeckesser replied:
Actually, all the non-pip installs should fail, because travis-test.sh doesn't explicitly install cython.  The pip installs should work, because pip reads pyproject.toml and does a "private" install of the build requires that it find there.

WarrenWeckesser replied:
The two jobs that set `USE_WHEEL=1` and the one that sets `USE_DEBUG=1` fail to build numpy, because they build numpy by running `setup.py`, but cython is not installed.

mattip replied:
Maybe add that to the comment: `version requirement stored there.` -> `version requirement for the USE_WHEEL=1 and USE_DEBUG=1 builds.`

WarrenWeckesser replied:
Yeah, I can change that comment to be clearer.

The *version requirement* is for any build of numpy, not just those that set `USE_WHEEL` or `USE_DEBUG`.  I put that comment there as an attempt to explain why I get the cython package argument from `test_requirements.txt` instead of just writing

    pip install setuptools wheel cython

I'll work on better wording.

Even better would be to read the build requirements directly from pyproject.toml.  I had that in an earlier version of this PR, but I got the sense that it seemed too complicated.

(Unfortunately, we currently have the required version of cython hard-coded in three places: pyproject.toml, test_requirements.txt, and tools/cythonize.py.  Not sure it is worth the effort at the moment to fix that; even if it is, it can probably wait for a different PR.)

WarrenWeckesser replied:
I updated the comments.  Tests are running again...

WarrenWeckesser commented:
Slightly on-topic, if the topic is "clean up the travis-ci shell scripts"...

Currently the last few lines of `.travis.yml` are

    before_install:
      - ./tools/travis-before-install.sh

    script:
      - ./tools/travis-test.sh

    after_success:
      - ./tools/travis-upload-wheel.sh


The *only* thing we do in the `before_install` section is run a shell script, and then we run another shell script in the `script` section.  As far as I can tell, there is no need to have two scripts like this.  We could move the contents of `travis-before-install.sh` into the file `travis-test.sh` and eliminate the `before-install` section.

For now, this is just an observation.  I'm not going to make that change here, unless someone else agrees with that assessment and thinks it is worthwhile.

mattip commented:
Thanks @WarrenWeckesser
